avro/src/main/java/com/fasterxml/jackson/dataformat/avro/AvroMapper.java

```diff
@@ -126,4 +146,85 @@ public AvroSchema schemaFrom(File schemaFile) throws IOException
                 .parse(schemaFile));
     }
+
+    public static class AvroSchemaTypeResolverBuilder extends StdTypeResolverBuilder {
+
+        @Override
+        public TypeSerializer buildTypeSerializer(SerializationConfig config, JavaType baseType, Collection<NamedType> subtypes) {
+            // No default typing for Serialization - It's already encoded in the Schema object
+            return null;
+        }
+
+        @Override
+        public TypeDeserializer buildTypeDeserializer(DeserializationConfig config, JavaType baseType, Collection<NamedType> subtypes) {
+            if (subtypes == null) {
+                BeanDescription bean = config.introspectClassAnnotations(baseType.getRawClass());
+                AnnotatedClass  ac   = bean.getClassInfo();
+                subtypes = config.getSubtypeResolver().collectAndResolveSubtypesByTypeId(config, ac);
+            }
+            return new AvroSchemaTypeDeserializer(
+                baseType,
+                new AvroNativeTypeIdResolver(baseType, config.getTypeFactory(), subtypes),
+                getDefaultImpl() != null ? config.getTypeFactory().constructType(getDefaultImpl()) : baseType
+            );
+        }
+    }
+
+    public static class AvroSchemaTypeDeserializer extends TypeDeserializerBase {
+        protected AvroSchemaTypeDeserializer(JavaType baseType, TypeIdResolver idRes, JavaType defaultImpl) {
+            super(baseType, idRes, null, true, defaultImpl);
+        }
+
+        protected AvroSchemaTypeDeserializer(TypeDeserializerBase src, BeanProperty property) {
+            super(src, property);
+        }
+
+        @Override
+        public Object deserializeTypedFromObject(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromArray(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromScalar(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromAny(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            // Primitives have no subtypes. Whatever the base type is, is correct!
+            if (!_baseType.isPrimitive() && jsonParser.canReadTypeId()) {
+                Object typeId = jsonParser.getTypeId();
+                if (typeId != null) {
+                    try {
+                        return _deserializeWithNativeTypeId(jsonParser, context, typeId);
+                    } catch (InvalidTypeIdException e) {
+                        // If the base type is acceptable, fall through; otherwise, propogate error
```

Comment:
This was done to handle the case when Jackson had type information, but the type information in the schema was invalid (for example, if data serialized from type A was being deserialized into type B, which is not a subtype of A or A is not even a class that exists on the class path), then we fall back to using the type Jackson has.

I added the code that queries the subtype resolver after I wrote this, so it could in theory be removed and instead devs would be required to register explicit named subtypes to map these invalid schema type IDs to valid classes on the class path instead, this just seemed like a nicer developer experience when your type names didn't match up 100% for some reason with your POJOs.